beacon-chain/rpc/beacon/attestations.go

```diff
@@ -45,53 +45,34 @@ func (bs *Server) ListAttestations(
 		return nil, status.Errorf(codes.InvalidArgument, "Requested page size %d can not be greater than max size %d",
 			req.PageSize, flags.Get().MaxPageSize)
 	}
-	var atts []*ethpb.Attestation
+	var blocks []*ethpb.SignedBeaconBlock
 	var err error
 	switch q := req.QueryFilter.(type) {
-	case *ethpb.ListAttestationsRequest_Genesis:
-		genBlk, err := bs.BeaconDB.GenesisBlock(ctx)
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "Could not genesis block: %v", err)
-		}
-		if genBlk == nil {
-			return nil, status.Error(codes.Internal, "Could not find genesis block")
-		}
-		genesisRoot, err := ssz.HashTreeRoot(genBlk.Block)
-		if err != nil {
-			return nil, err
-		}
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetHeadBlockRoot(genesisRoot[:]))
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "Could not fetch genesis attestations: %v", err)
-		}
-	case *ethpb.ListAttestationsRequest_HeadBlockRoot:
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetHeadBlockRoot(q.HeadBlockRoot))
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
-		}
-	case *ethpb.ListAttestationsRequest_SourceEpoch:
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetSourceEpoch(q.SourceEpoch))
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
-		}
-	case *ethpb.ListAttestationsRequest_SourceRoot:
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetSourceRoot(q.SourceRoot))
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
-		}
-	case *ethpb.ListAttestationsRequest_TargetEpoch:
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetTargetEpoch(q.TargetEpoch))
+	case *ethpb.ListAttestationsRequest_GenesisEpoch:
+		blocks, err = bs.BeaconDB.Blocks(ctx, filters.NewFilter().SetStartEpoch(0).SetEndEpoch(0))
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
 		}
-	case *ethpb.ListAttestationsRequest_TargetRoot:
-		atts, err = bs.BeaconDB.Attestations(ctx, filters.NewFilter().SetTargetRoot(q.TargetRoot))
+	case *ethpb.ListAttestationsRequest_Epoch:
+		blocks, err = bs.BeaconDB.Blocks(ctx, filters.NewFilter().SetStartEpoch(q.Epoch).SetEndEpoch(q.Epoch))
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "Could not fetch attestations: %v", err)
 		}
 	default:
 		return nil, status.Error(codes.InvalidArgument, "Must specify a filter criteria for fetching attestations")
 	}
+	totalLen := 0
+	for _, block := range blocks {
+		totalLen += len(block.Block.Body.Attestations)
+	}
+	atts := make([]*ethpb.Attestation, totalLen)
```

Comment:
it might be better to just do appends but just set the capacity to slots per epoch * max attestations rather than looping over blocks twice

Reply:
although i don't have concrete proof of it

Reply:
we can try a simple benchmark for this